File: js/protocols/RoutingEngine.js
(function(exports) {
    const NetworkInterface = (typeof require !== 'undefined') ? require('../models/NetworkInterface').NetworkInterface : exports.NetworkInterface;
    const RouteEntry = (typeof require !== 'undefined') ? require('../models/RouterDevice').RouteEntry : exports.RouteEntry;
    const RouterDevice = (typeof require !== 'undefined') ? require('../models/RouterDevice').RouterDevice : exports.RouterDevice;

    class RoutingEngine {
        static computeAllRoutes(devices, links) {
            for (const dev of devices) {
                if (dev instanceof RouterDevice || dev.getDeviceType() === 'L3 Router') {
                    RoutingEngine.computeRoutesForRouter(dev, devices, links);
                }
            }
        }

        static computeRoutesForRouter(router, allDevices, allLinks) {
            router.clearRoutes();

            // 1. Directly Connected Networks
            for (const iface of router.interfaces) {
                if (iface.isConfigured() && iface.isUp) {
                    const network = NetworkInterface.getNetworkAddress(iface.ipAddress, iface.subnetMask);
                    router.addRoute(new RouteEntry(network, iface.subnetMask, 'Direct', iface.name, 0, 'DIRECT'));
                }
            }

            // 2. Dijkstra SPF Algorithm across the topology graph
            const distances = new Map();
            const firstInterface = new Map();
            const unvisited = new Set();

            for (const d of allDevices) {
                distances.set(d, Infinity);
                unvisited.add(d);
            }

            distances.set(router, 0);

            while (unvisited.size > 0) {
                // Find node in unvisited with smallest distance
                let current = null;
                let minDist = Infinity;
                for (const node of unvisited) {
                    const d = distances.get(node);
                    if (d < minDist) {
                        minDist = d;
                        current = node;
                    }
                }

                if (!current || minDist === Infinity) break;
                unvisited.delete(current);

                // Explore neighbors
                for (const uIface of current.interfaces) {
                    if (!uIface.isUp) continue;
                    const link = uIface.connectedLink;
                    if (!link || !link.isUp) continue;

                    const v = link.getOtherDevice(current);
                    if (!v || !v.isOnline || !unvisited.has(v)) continue;

                    const linkCost = link.getRoutingCost();
                    const newDist = distances.get(current) + linkCost;

                    if (newDist < distances.get(v)) {
                        distances.set(v, newDist);

                        if (current === router) {
                            firstInterface.set(v, uIface);
                        } else {
                            firstInterface.set(v, firstInterface.get(current));
                        }
                    }
                }
            }

            // 3. Populate routes for all reachable subnets
            for (const target of allDevices) {
                if (target === router || distances.get(target) === Infinity) continue;

                const outIface = firstInterface.get(target);
                if (!outIface) continue;

                const metricCost = Math.round(distances.get(target));
                let nextHopIp = 'Direct';
                const link = outIface.connectedLink;
                if (link) {
                    const otherIface = link.getOtherInterface(outIface);
                    if (otherIface && otherIface.ipAddress) {
                        nextHopIp = otherIface.ipAddress;
                    }
                }

                for (const targetIface of target.interfaces) {
                    if (targetIface.isConfigured()) {
                        const targetNet = NetworkInterface.getNetworkAddress(targetIface.ipAddress, targetIface.subnetMask);
                        const exists = router.routingTable.some(r =>
                            r.destinationNetwork === targetNet && r.subnetMask === targetIface.subnetMask
                        );

                        if (!exists) {
                            router.addRoute(new RouteEntry(
                                targetNet,
                                targetIface.subnetMask,
                                nextHopIp,
                                outIface.name,
                                metricCost,
                                'OSPF'
                            ));
                        }
                    }
                }
            }
        }
    }

    exports.RoutingEngine = RoutingEngine;
})(typeof module !== 'undefined' && module.exports ? module.exports : (window.CampusNet = window.CampusNet || {}));

File: src/com/campusnet/protocol/RoutingEngine.java
package com.campusnet.protocol;

import com.campusnet.model.*;

import java.util.*;

public class RoutingEngine {

    public static void computeAllRoutes(List<Device> devices, List<NetworkLink> links) {
        for (Device dev : devices) {
            if (dev instanceof RouterDevice router) {
                computeRoutesForRouter(router, devices, links);
            }
        }
    }

    public static void computeRoutesForRouter(RouterDevice router, List<Device> allDevices, List<NetworkLink> allLinks) {
        router.clearRoutes();

        // 1. Add Directly Connected Networks
        for (NetworkInterface iface : router.getInterfaces()) {
            if (iface.isConfigured() && iface.isUp()) {
                String network = getNetworkAddress(iface.getIpAddress(), iface.getSubnetMask());
                router.addRoute(new RouteEntry(network, iface.getSubnetMask(), "Direct", iface.getName(), 0, "DIRECT"));
            }
        }

        // 2. Dijkstra Shortest Path Algorithm across the network graph
        Map<Device, Double> distances = new HashMap<>();
        Map<Device, Device> previousHop = new HashMap<>();
        Map<Device, NetworkInterface> firstInterface = new HashMap<>();
        PriorityQueue<NodeDistance> pq = new PriorityQueue<>(Comparator.comparingDouble(a -> a.distance));

        for (Device d : allDevices) {
            distances.put(d, Double.POSITIVE_INFINITY);
        }

        distances.put(router, 0.0);
        pq.add(new NodeDistance(router, 0.0));

        while (!pq.isEmpty()) {
            NodeDistance current = pq.poll();
            Device u = current.device;

            if (current.distance > distances.get(u)) continue;

            // Explore neighbors
            for (NetworkInterface uIface : u.getInterfaces()) {
                if (!uIface.isUp()) continue;
                NetworkLink link = uIface.getConnectedLink();
                if (link == null || !link.isUp()) continue;

                Device v = link.getOtherDevice(u);
                if (v == null || !v.isOnline()) continue;

                double linkCost = link.getRoutingCost();
                double newDist = distances.get(u) + linkCost;

                if (newDist < distances.get(v)) {
                    distances.put(v, newDist);
                    previousHop.put(v, u);

                    if (u == router) {
                        firstInterface.put(v, uIface);
                    } else {
                        firstInterface.put(v, firstInterface.get(u));
                    }

                    pq.add(new NodeDistance(v, newDist));
                }
            }
        }

        // 3. For all reachable devices and their subnets, populate the routing table
        for (Device target : allDevices) {
            if (target == router || distances.get(target) == Double.POSITIVE_INFINITY) continue;

            NetworkInterface outIface = firstInterface.get(target);
            if (outIface == null) continue;

            int metricCost = (int) Math.round(distances.get(target));

            // Find next hop IP on outIface
            String nextHopIp = "Direct";
            NetworkLink link = outIface.getConnectedLink();
            if (link != null) {
                NetworkInterface otherIface = link.getOtherInterface(outIface);
                if (otherIface != null && otherIface.getIpAddress() != null) {
                    nextHopIp = otherIface.getIpAddress();
                }
            }

            for (NetworkInterface targetIface : target.getInterfaces()) {
                if (targetIface.isConfigured()) {
                    String targetNet = getNetworkAddress(targetIface.getIpAddress(), targetIface.getSubnetMask());
                    // Check if already in routing table
                    boolean exists = false;
                    for (RouteEntry re : router.getRoutingTable()) {
                        if (re.getDestinationNetwork().equals(targetNet) && re.getSubnetMask().equals(targetIface.getSubnetMask())) {
                            exists = true;
                            break;
                        }
                    }
                    if (!exists) {
                        router.addRoute(new RouteEntry(
                                targetNet,
                                targetIface.getSubnetMask(),
                                nextHopIp,
                                outIface.getName(),
                                metricCost,
                                "OSPF"
                        ));
                    }
                }
            }
        }
    }

    public static String getNetworkAddress(String ip, String mask) {
        if (ip == null || mask == null) return "0.0.0.0";
        try {
            long ipVal = NetworkInterface.ipToLong(ip);
            long maskVal = NetworkInterface.ipToLong(mask);
            long netVal = ipVal & maskVal;
            return NetworkInterface.longToIp(netVal);
        } catch (Exception e) {
            return "0.0.0.0";
        }
    }

    private static class NodeDistance {
        final Device device;
        final double distance;

        NodeDistance(Device device, double distance) {
            this.device = device;
            this.distance = distance;
        }
    }
}
